src/components/widget/MultiSelect.js

```diff
@@ -0,0 +1,112 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+
+class MultiSelect extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      data: this.props.options,
+      filter: '',
+      checkedItems: {},
+    };
+  }
+
+  componentDidMount() {
+    this.props.onFocus();
+  }
+
+  static getDerivedStateFromProps(nextProps) {
+    let selected =
+      nextProps.selectedItems && Array.isArray(nextProps.selectedItems)
+        ? nextProps.selectedItems
+        : nextProps.selectedItems !== null
+        ? nextProps.selectedItems.values
+        : null;
+    if (selected !== null) {
+      let updatedCheckedItems = {};
+      selected.map((item) => {
+        if (item) {
+          updatedCheckedItems[item.key] = {
+            key: item.key,
+            caption: item.caption,
+            value: true,
+          };
+        }
+        return item;
+      });
+      return {
+        checkedItems: updatedCheckedItems,
+      };
+    }
+
+    return null;
+  }
+
+  selectItem = (key, caption) => {
+    let selected = null;
+    let newCheckedItems = JSON.parse(JSON.stringify(this.state.checkedItems));
+    if (typeof this.state.checkedItems[key] === 'undefined') {
+      newCheckedItems[key] = { key: key, caption: caption, value: true };
+    } else {
+      newCheckedItems[key].value = !this.state.checkedItems[key].value;
+    }
+    this.setState({ checkedItems: newCheckedItems });
+    selected = Object.keys(newCheckedItems).map((item) => {
+      if (newCheckedItems[item].value === true) {
+        return {
+          key: newCheckedItems[item].key,
+          caption: newCheckedItems[item].caption,
+        };
+      }
+    });
+
+    this.props.onSelect(selected.filter((el) => typeof el !== 'undefined'));
```

Comment:
Just wondering if we shouldn't call this after we've updated the state ? Or it won't make a difference here ?

Reply:
calling it here will guarantee the top filter is set accordingly. The sync to state is done in getDerivedStateFromProps